File: [0094][Medium][Binary_Tree_Inorder_Traversal][BST]/Binary_Tree_Inorder_Traversal.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 56 ms
# leetcode memory   cost : 16.2 MB
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# python 中序遍历

class Solution(object):
    def __init__(self):
        self.travelpath = []
 
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
    
        if root is not None:
            
            self.inorderTraversal(root.left)
            self.travelpath.append(root.val)
            self.inorderTraversal(root.right)
        return self.travelpath  

File: [0094][Medium][Binary_Tree_Inorder_Traversal][BST]/Binary_Tree_Inorder_Traversal.java
//Definition for a binary tree node.
public class TreeNode {
     int val;
    TreeNode left;
     TreeNode right;
     TreeNode(int x) { val = x; }
}

// Time  Complexity: O(N)
// Space Complexity: O(N)
// Solution 1, recursion
class Solution {
    public List < Integer > inorderTraversal(TreeNode root) {
        List < Integer > res = new ArrayList < > ();
        helper(root, res);
        return res;
    }

    public void helper(TreeNode root, List < Integer > res) {
        if (root != null) {
            if (root.left != null) {
                helper(root.left, res);
            }
            res.add(root.val);
            if (root.right != null) {
                helper(root.right, res);
            }
        }
    }
}